src/com/facebook/buck/jvm/java/JavaSourceJar.java

```diff
@@ -81,12 +99,21 @@ public JavaSourceJar(
     steps.add(new MakeCleanDirectoryStep(getProjectFilesystem(), temp));
 
     Set<Path> seenPackages = Sets.newHashSet();
+    Map<ProjectFilesystem, JavaPackageFinder> finders = new HashMap<>();
+
+    for (SourcePath sourcePath : sources) {
+      Path source = getResolver().getAbsolutePath(sourcePath);
+      if (!"java".equals(MorePaths.getFileExtension(source))) {
```

Comment:
This isn't a full blown java parser, it just handles enough to get to the package declaration. Parsing files using regex is generally considered a pretty [bad idea](https://blog.codinghorror.com/parsing-beyond-regex/), and possibly error-prone since there might be a package declaration in a comment. Put another way, we either need a very complex regex or a simple parser in order to be consistently correct. I opted for a parser.

Given that this parser is only used from the `JavaSourceJar` class, I'm not sure why you mention kotlin, Scala, and groovy. It seems beside the point of this diff.